### PR TITLE
Add shortcuts for refs.pow

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81113
* #78616
* #78615
* #78350
* __->__ #80219

As per title

cc @ezyang @mruberry @ngimel @Lezcano @peterbell10

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -822,6 +822,33 @@ def add(
...
+
+    # If b is a number. mypy doesn't like isinstance(b, NumberType)
+    if not isinstance(b, TensorLikeType):
+        if b == 1.0:
+            return a  # type: ignore[return-value]
+        elif b == 2.0:
+            return a * a  # type: ignore[return-value]
+        elif b == 0.5:
+            return torch.sqrt(a)  # type: ignore[arg-type]
+        elif b == 1.0 / 3.0:
```
Comment (mruberry) [inline torch/_refs/__init__.py:840]: Nice addition

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -852,7 +879,7 @@ def float_power(
         b = prims.to_dtype(b, dtype)
 
     a, b = _maybe_broadcast(a, b)
-    return prims.pow(a, b)
+    return pow(a, b)
```
Comment (mruberry) [inline torch/_refs/__init__.py:882]: `a**b`

**Comment by mruberry** [comment]
Hey @Lezcano! Super cool, but looks like this will need to update some test annotations

**Comment by Lezcano** [comment]
The only test that seems related is https://github.com/pytorch/pytorch/runs/7062796377?check_suite_focus=true, but I do not know the reason for its fail or how to resolve it. Do you know what's going on there?

**Comment by mruberry** [comment]
> The only test that seems related is https://github.com/pytorch/pytorch/runs/7062796377?check_suite_focus=true, but I do not know the reason for its fail or how to resolve it. Do you know what's going on there?

Maybe an error in nvFuser or the bindings -- cc @IvanYashchuk. I would just xfail the test for now with a comment about what you're seeing

File: torch/_refs/__init__.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:851]: But unfortunately incorrect when `a` is negative.

For mathematical details see: https://www.intmath.com/blog/mathematics/is-cube-root-the-same-as-raising-to-power-13-10460

There's no `torch.cbrt` and `prims.cbrt` is currently implemented as:
```
def _cbrt_aten(a: torch.Tensor):
    return pow(a, (1 / 3))
```
which is also incorrect for negative values and doesn't match NumPy or C.

The correct implementation for the `cbrt` function in terms of `pow` would be
```py
def cbrt(x):
    return -((-x)**(1/3)) if x < 0 else x**(1/3)
```

```py
def cbrt(x):
    return -((-x)**(1/3)) if x < 0 else x**(1/3)
    
import numpy as np
a = -2.0
print(np.cbrt(a)) # -1.259921049894873
print(a ** (1 / 3)) # (0.6299605249474367+1.0911236359717214j)
print(cbrt(a)) # -1.2599210498948732
print(np.power(a, 1 / 3)) # nan

import torch
print(torch.tensor(a), 1 / 3)) # tensor(nan)
```

In Python 3.11 `cbrt` is added to the `math` library https://github.com/python/cpython/commit/ac867f10b49322e25f34d2d8abd8e63c86834750.

**Comment by IvanYashchuk** [comment]
> > The only test that seems related is https://github.com/pytorch/pytorch/runs/7062796377?check_suite_focus=true, but I do not know the reason for its fail or how to resolve it. Do you know what's going on there?
> 
> Maybe an error in nvFuser or the bindings -- cc @IvanYashchuk. I would just xfail the test for now with a comment about what you're seeing

It should be fine to xfail the test, but in this case, I can't even run it on this PR rebased on the latest master, `python -m pytest test/test_ops.py -k "test_python_ref_executor__refs_pow_executor_nvfuser_cuda_int32"` gives me:
```py
collected 0 items / 1 error

========================================================================== ERRORS ==========================================================================
____________________________________________________________ ERROR collecting test/test_ops.py _____________________________________________________________
Traceback (most recent call last):
  File "/home/iyashchuk/dev/pytorch/master/test/test_ops.py", line 8, in <module>
    import torch
  File "/home/iyashchuk/dev/pytorch/master/torch/__init__.py", line 968, in <module>
    from . import _meta_registrations
  File "/home/iyashchuk/dev/pytorch/master/torch/_meta_registrations.py", line 568, in <module>
    import torch._refs
  File "/home/iyashchuk/dev/pytorch/master/torch/_refs/__init__.py", line 863, in <module>
    def div(
  File "/home/iyashchuk/dev/pytorch/master/torch/_prims/wrappers.py", line 174, in out_wrapper
    assert is_tensor or len(out_names) >= 2
AssertionError
================================================================= short test summary info ==================================================================
ERROR test/test_ops.py - AssertionError
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
===================================================================== 1 error in 0.83s =====================================================================
```

Could you please rebase first?

File: torch/_refs/__init__.py (hunk 1 above)
Comment (Lezcano) [reply torch/_refs/__init__.py:851]: wtf.

From [`np.power`](https://numpy.org/doc/stable/reference/generated/numpy.power.html):
>Negative values raised to a non-integral value will return nan. To get complex results, cast the input to complex, or specify the dtype to be complex (see the example below).

Inherited from [`std::pow`](https://en.cppreference.com/w/cpp/numeric/math/pow):
> pow(base, exp) returns NaN and raises [FE_INVALID](https://en.cppreference.com/w/cpp/numeric/fenv/FE_exceptions) if base is finite and negative and exp is finite and non-integer.

Now, after thinking about this for quite some time, this behaviour, although very counterintuitive, makes sense. 

tl;dr. `cbrt` is defined for real numbers as "the usual cubic root", while `pow(a, z)` is defined as `exp(zlog(a))` where `log(a)` represents the main branch of the complex logarithm. The function `z^{1/3}` has 3 possible values that are always arranged on a triangle of the form:
image
The definition `exp(zlog(a))` happens to be choosing the first root (counterclockwise) in that diagram starting at 1. This happens for all negative numbers, and it's annoying, as the real root of `z^{1/3}` is the second one. Note that, for higher roots like `z^{1/4}` or `z^{1/5}` we get similar diagrams to this one but with a square, a pentagon, etc. In general, you'd need to choose a very particular branch of the logarithm to get the real root, which would make for a very odd casing for negative numbers in `std::pow`. As such, the  C++ committee settled for this behaviour which, although weird, is correct and efficient.

So, from here we get a few things:
- I'll remove that `cbrt` from there
- We need to fix `cbrt`
- We need to give good definitions for what we mean with `sqrt` (this one's easy), `pow` and `cbrt`.

**Comment by Lezcano** [comment]
@IvanYashchuk sorry, I forgot to update a small thing in the first PR of this stack. Now it should be fine.

File: torch/_refs/__init__.py (hunk 1 above)
Comment (Lezcano) [reply torch/_refs/__init__.py]: fwiw, I summarised these points in a small comment in https://github.com/pytorch/pytorch/pull/80443. Hope that works. Thank you for pointing this out @IvanYashchuk!

File: torch/_refs/__init__.py (hunk 2 above)
Comment (Lezcano) [reply torch/_refs/__init__.py:947]: This was failing with some error. Leaving it as `pow(a, b)`.

Comment (Lezcano) [reply torch/_refs/__init__.py:947]: fwiw it throws a `RuntimeError: no _refs support for torch.Tensor.__pow__`

**Review by mruberry — APPROVED** [review]
Cool!

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -923,6 +923,30 @@ def div(
...
+
+def _pow(
+    a: Union[TensorLikeType, NumberType],
+    b: Union[TensorLikeType, NumberType],
+) -> TensorLikeType:
+    assert isinstance(a, TensorLikeType) or isinstance(b, TensorLikeType)
+
+    if isinstance(b, Number):
+        if b == 1.0:
+            return a  # type: ignore[return-value]
```
Comment (peterbell10) [inline torch/_refs/__init__.py:935]: It seems dangerous to sometimes return a new tensor and sometimes not
```suggestion
            return a.clone()  # type: ignore[return-value]
```

Comment (mruberry) [reply torch/_refs/__init__.py:935]: Good catch, @peterbell10! `torch.pow` always returns a new tensor
